src/compiler/java_compiler.nit

```diff
@@ -0,0 +1,2297 @@
+				var rt_name = b.mclass.rt_name
+				s.add "({value1}.rtclass == {rt_name}.get{rt_name}()) && ({value1}.value.equals({value2}.value))"
+				if b.mclass.name == "Float" then
+					s.add "({value1}.rtclass == RTClass_kernel_Float.getRTClass_kernel_Float() && {value1}.rtclass == {value2}.rtclass && Math.abs((double)({value1}.value)) == 0.0 && Math.abs((double)({value2}.value)) == 0.0)"
+				end
+			end
+			if s.is_empty then
+				self.add("{res} = {value1} == {value2}; /* both can be primitive */")
+				return res
+			end
+			test.add("({s.join(" || ")})")
+		else
+			self.add("{res} = {value1} == {value2}; /* no primitives */")
+			return res
+		end
+		self.add("{res} = {value1} == {value2} || ({test.join(" && ")});")
+		return res
+	end
+
+	# Attributes
+
+	# Generate a polymorphic attribute is_set test
+	fun isset_attribute(a: MAttribute, recv: RuntimeVariable): RuntimeVariable do
+		# TODO self.check_recv_notnull(recv)
+		var res = new_var(compiler.mainmodule.bool_type)
+
+		# What is the declared type of the attribute?
+		var mtype = a.intro.static_mtype.as(not null)
+		var intromclassdef = a.intro.mclassdef
+		mtype = mtype.resolve_for(intromclassdef.bound_mtype, intromclassdef.bound_mtype, intromclassdef.mmodule, true)
+
+		if mtype isa MNullableType then
+			add("{res} = true; /* easy isset: {a} on {recv.inspect} */")
+			return res
+		end
+		add("{res} = {recv}.attrs.get(\"{a.jname}\") != null; /* {a} on {recv.inspect} */")
+		return res
+	end
+
+	# Generate a polymorphic attribute read
+	fun read_attribute(a: MAttribute, recv: RuntimeVariable): RuntimeVariable do
+		# TODO check_recv_notnull(recv)
+		# TODO compile_check(v)
+		# What is the declared type of the attribute?
+		var ret = a.intro.static_mtype.as(not null)
+		var intromclassdef = a.intro.mclassdef
+		ret = ret.resolve_for(intromclassdef.bound_mtype, intromclassdef.bound_mtype, intromclassdef.mmodule, true)
+
+		# Check for Uninitialized attribute
+		if not ret isa MNullableType then check_attribute(a, recv)
+
+		return new_expr("{recv}.attrs.get(\"{a.jname}\")", ret)
+	end
+
+	# Generate a polymorphic attribute write
+	fun write_attribute(a: MAttribute, recv: RuntimeVariable, value: RuntimeVariable) do
+		# TODO check_recv_notnull(recv)
+		add "{recv}.attrs.put(\"{a.jname}\", {autobox(value, compiler.mainmodule.object_type)});"
+	end
+
+	# Check uninitialized attribute
+	fun check_attribute(a: MAttribute, recv: RuntimeVariable) do
+		add "if({recv}.attrs.get(\"{a.jname}\") == null) \{"
+		add_abort "Uninitialized attribute {a.name}"
+		add "\}"
+	end
+
+	# Utils
+
+	# Display a info message
+	fun info(str: String) do compiler.modelbuilder.toolcontext.info(str, 0)
+end
+
+# A file containing Java code.
+class JavaCodeFile
+
+	# File name
+	var filename: String
+
+	# Lines to write
+	var lines: List[String] = new List[String]
+end
+
+redef class MEntity
+	# A Java compatible name for `self`
+	private fun jname: String do return name.to_cmangle
```

Comment:
warning: nothing prevent collision of names. Unless you do not care about collision you should use `MEntity::c_name` instead